tools/shared.py

```diff
@@ -2335,6 +2341,37 @@ class JS(object):
   def to_nice_ident(ident): # limited version of the JS function toNiceIdent
     return ident.replace('%', '$').replace('@', '_').replace('.', '_')
 
+  # Returns the subresource location for run-time access
+  @staticmethod
+  def get_subresource_location(path, data_uri=None):
+    if data_uri is None:
+      data_uri = Settings.SINGLE_FILE
+    if data_uri:
+      f = open(path, 'rb')
+      data = base64.b64encode(f.read())
+      f.close()
+      return 'data:application/octet-stream;base64,' + data
+    else:
+      return os.path.basename(path)
+
+  # Runs the preprocessor on a string of JavaScript
+  #
+  # WARNING: Currently using a very simplified version that supports only a
+  # small subset of preprocessor directives. Specifically, it will work with
+  # "#if NAME_OF_SETTING ... #endif". Anything other than that, including
+  # "#if !NAME_OF_SETTING ... #endif", is unsupported.
+  #
+  # TODO: Replace this with a hook into the real preprocessor.
+  @staticmethod
+  def preprocessor(s):
+    settings = settings or Settings
```

Comment:
this looks bad, where would a global `settings` object come from? we should not depend on one. can't you just use `Settings`? or maybe it needs settings as a parameter.